go/adbc/driver/snowflake/binding.go

```diff
@@ -0,0 +1,141 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package snowflake
+
+import (
+	"database/sql"
+	"database/sql/driver"
+	"fmt"
+
+	"github.com/apache/arrow-adbc/go/adbc"
+	"github.com/apache/arrow/go/v17/arrow"
+	"github.com/apache/arrow/go/v17/arrow/array"
+)
+
+func convertArrowToNamedValue(batch arrow.Record, index int) ([]driver.NamedValue, error) {
+	// see goTypeToSnowflake in gosnowflake
+	// technically, snowflake can bind an array of values at once, but
+	// only for INSERT, so we can't take advantage of that without
+	// analyzing the query ourselves
+	params := make([]driver.NamedValue, batch.NumCols())
+	for i, field := range batch.Schema().Fields() {
+		rawColumn := batch.Column(i)
+		params[i].Ordinal = i + 1
+		switch column := rawColumn.(type) {
+		case *array.Boolean:
+			params[i].Value = sql.NullBool{
+				Bool:  column.Value(index),
+				Valid: column.IsValid(index),
+			}
+		case *array.Float32:
+			// Snowflake only recognizes float64
+			params[i].Value = sql.NullFloat64{
+				Float64: float64(column.Value(index)),
+				Valid:   column.IsValid(index),
+			}
+		case *array.Float64:
+			params[i].Value = sql.NullFloat64{
+				Float64: column.Value(index),
+				Valid:   column.IsValid(index),
+			}
+		case *array.Int8:
+			// Snowflake only recognizes int64
+			params[i].Value = sql.NullInt64{
+				Int64: int64(column.Value(index)),
+				Valid: column.IsValid(index),
+			}
+		case *array.Int16:
+			params[i].Value = sql.NullInt64{
+				Int64: int64(column.Value(index)),
+				Valid: column.IsValid(index),
+			}
+		case *array.Int32:
+			params[i].Value = sql.NullInt64{
+				Int64: int64(column.Value(index)),
+				Valid: column.IsValid(index),
+			}
+		case *array.Int64:
+			params[i].Value = sql.NullInt64{
+				Int64: column.Value(index),
+				Valid: column.IsValid(index),
+			}
+		case *array.String:
+			params[i].Value = sql.NullString{
+				String: column.Value(index),
+				Valid:  column.IsValid(index),
+			}
+		case *array.LargeString:
+			params[i].Value = sql.NullString{
+				String: column.Value(index),
+				Valid:  column.IsValid(index),
+			}
+		default:
+			return nil, adbc.Error{
+				Code: adbc.StatusNotImplemented,
+				Msg:  fmt.Sprintf("[Snowflake] Unsupported bind param '%s' type %s", field.Name, field.Type.String()),
+			}
+		}
+	}
+	return params, nil
+}
+
+type snowflakeBindReader struct {
+	doQuery      func([]driver.NamedValue) (array.RecordReader, error)
+	currentBatch arrow.Record
+	nextIndex    int64
+	// may be nil if we bound only a batch
+	stream array.RecordReader
+}
+
+func (r *snowflakeBindReader) Release() {
+	if r.currentBatch != nil {
+		r.currentBatch.Release()
+	}
```

Comment:
we might still end up with a double release here because we don't set `r.currentBatch` to `nil` after calling `stream.Next()` and getting the end of the stream. we should probably just always set `r.currentBatch = nil` before we call `stream.Next()` to ensure we don't double release. You can test for excess releases by running with `-tags assert` which will turn on the debug asserts in Arrow